File: 프로그래머스/unrated/154538. 숫자 변환하기/숫자 변환하기.js
function solution(x, y, n) {
    if (x === y) return 0;

    const dp = Array.from({ length: y + 1 }).fill(Infinity);
    dp[x] = 0;

    for (let i = x + 1; i <= y; ++i) {
        if (i - n >= 0) dp[i] = dp[i - n] + 1;
        if (!(i % 2)) dp[i] = Math.min(dp[Math.floor(i / 2)] + 1, dp[i]);
        if (!(i % 3)) dp[i] = Math.min(dp[Math.floor(i / 3)] + 1, dp[i]);
    }

    return (dp[y] !== Infinity) ? dp[y] : -1;
}

File: 프로그래머스/unrated/154538. 숫자 변환하기/숫자 변환하기.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

void search(int x, int y, int n, int count, vector<int>& v)
{
    if (v[x] && count >= v[x]) return;
    
    v[x] = count;

    if (x + n <= y) search(x + n, y, n, count + 1, v);
    if (x * 2 <= y) search(x * 2, y, n, count + 1, v);
    if (x * 3 <= y) search(x * 3, y, n, count + 1, v);
}

int solution(int x, int y, int n) {
    if (x == y) return 0;
    
    vector<int> v(y, 0);
    search(x, y, n, 0, v);
    return !v[y] ? -1 : v[y];
}